### PR TITLE
[quant][graphmode][fx][fp16] Add fp16 support for layer_norm

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52865 [quant][graphmode][fx][fp16] Add fp16 support for silu
* #52864 [quant][graphmode][fx][fp16] Add fp16 support for tanh
* #52863 [quant][graphmode][fx][fp16] Add fp16 support for sigmoid
* **#52862 [quant][graphmode][fx][fp16] Add fp16 support for layer_norm**
* #52811 [quant][graphmode][fx][fp16] Add fp16 support for torch.sum
* #52810 [quant][graphmode][fx][fp16] Add fp16 support for div
* #52809 [quant][graphmode][fx][fp16] Add fp16 support for sub

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

### CODE REVIEW

File: test/quantization/test_quantize_fx.py [hunk 1]
```diff
@@ -2622,11 +2622,52 @@ def forward(self, input):
...
+                    self.op = float_op
+
+            def forward(self, input):
+                if self.is_module:
+                    return self.op(input)
+                else:
+                    args = [input]
+                    if not skip_op_arg_for_functional:
+                        args += op_args
+                    return self.op(*args)
```
Comment (vkuzo) [inline test/quantization/test_quantize_fx.py:2647]: optional: this seems hard to read, at least for me (it may be different for others). I wonder if there were two less complicated classes, maybe one for modules and one for functions, if it would be easier to read.  I personally would value readability more than code reuse in this context.  I'm flexible though.

File: torch/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -725,41 +732,78 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
...
-                load_arg(quantized=False)(node.kwargs))
+
+        all_dtypes = [
+            (torch.quint8, torch.qint8, None),
+            (torch.float16, torch.float16, None)
+        ]
+        int8_dtypes = [
+            (torch.quint8, torch.qint8, None)
+        ]
+        supported_dtypes = {
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:743]: optional for a future PR, would it make sense for these mappings to live somewhere else in a mapping only file?

**Review by vkuzo — APPROVED** [review]
lg!

File: torch/quantization/fx/quantization_patterns.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:743]: yeah, I plan to refactor this part a bit later, maybe we can put the type tuple in a function. for mappings I think it would make sense to keep them here for now since they are related to the patterns above (in register_patterns), but maybe we can include them in the patterns, will need to think about it more. let me know if you have any suggestions.

File: test/quantization/test_quantize_fx.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:2647]: sure, this is a little bit hard to read, I can split it to two modules
